acnportal/acnsim/base.py

```diff
@@ -90,45 +86,108 @@ def __repr__(self):
+        # The code here is from pandas 1.0.1, io.json.to_json(), with
+        # modifications.
+        json_serializable_data = self._to_registry()[0]
+        if json_serializable_data["version"] is None:
+            warnings.warn(
+                f"Missing a recorded version of acnportal in the dict "
+                f"representation. Loading will not run an acnportal "
+                f"version check.",
+                UserWarning
+            )
+        if json_serializable_data["dependency_versions"] is None:
+            warnings.warn(
+                f"Missing recorded versions of dependencies in the "
+                f"dict representation. Loading will not run a "
+                f"dependency version check.",
+                UserWarning
+            )
+        path_or_buf = stringify_path(path_or_buf)
+        if isinstance(path_or_buf, str):
+            fh, _ = get_handle(path_or_buf, "w")
+            try:
+                json.dump(json_serializable_data, fh)
+                # Add a newline to the EOF.
+                fh.write("\n")
+            finally:
+                fh.close()
+        elif path_or_buf is None:
+            return json.dumps(json_serializable_data)
+        else:
+            json.dump(json_serializable_data, path_or_buf)
+            # Add a newline to the EOF.
+            path_or_buf.write("\n")
+
+    def _to_registry(self, context_dict=None):
         """
         Returns a JSON serializable representation of self.
 
         The serializer is invoked using `obj.to_json()`, but the
         conversion into a serializable representation occurs in this
         method.
 
+        This method is protected (i.e. underscored) so users do not
+        directly use this method. PyNoInspection comments have been
+        added where this function is used within the package.
+
         Serializer behaviors:
 
         - Any native ACN-Sim object (that is, any unaltered object
         provided in `acnsim`) dumps without error and with
         accurate preservation of attributes.
 
         - An extension of an ACN-Sim object that appropriately defines a
-        `to_dict` method will dump without error and with
+        `_to_dict` method will dump without error and with
         accurate attribute preservation. This also applies to ACN-Sim
         objects that have as attributes extensions of ACN-Sim objects
-        with defined `to_dict` methods.
+        with defined `_to_dict` methods.
 
-        - An extension that doesn't define `to_dict` will dump without
+        - An extension that doesn't define `_to_dict` will dump without
         error, but with partial preservation of attributes. Objects that
-        are not JSON serializable and which do not have a `to_registry`
+        are not JSON serializable and which do not have a `_to_registry`
         method will not be completely serialized unless said objects are
         extensions of built-in ACN-Sim objects with no extra attributes.
 
```

Comment:
On second thought, I think it would be best to say that all additional objects a user may wish to serialize must both inherit from a BaseSimObj-like and define a `to_dict` method. I originally said `to_registry` as I anticipated users creating objects that they wanted to serialize without inheriting from BaseSimObj. However, _to_registry contains so much overhead that is _expected_ by the deserializer that I think it would be easier to just recommend some workarounds, such as a wrapper class inheriting from BaseSimObj, or multiple inheritance. I can provide an example in a serializer tutorial, which is on my TODO list.